ethdb/bitmapdb/dbutils.go

```diff
@@ -3,121 +3,152 @@ package bitmapdb
 	"github.com/ledgerwatch/turbo-geth/ethdb"
 )
 
-const ColdShardLimit = 512 * datasize.KB
-const HotShardLimit = 4 * datasize.KB
+const ShardLimit = 3 * datasize.KB
+
+var blockNBytes = make([]byte, 4)
 
 // AppendMergeByOr - appending delta to existing data in db, merge by Or
 // Method maintains sharding - because some bitmaps are >1Mb and when new incoming blocks process it
 //	 updates ~300 of bitmaps - by append small amount new values. It cause much big writes (LMDB does copy-on-write).
 //
-// 3 terms are used: cold_shard, hot_shard, delta
-//   delta - most recent changes (appendable)
-//   hot_shard - merge delta here until hot_shard size < HotShardLimit, otherwise merge hot to cold
-//   cold_shard - merge hot_shard here until cold_shard size < ColdShardLimit, otherwise mark hot as cold, create new hot from delta
-// if no cold shard, create new hot from delta - it will turn hot to cold automatically
-// never merges cold with cold for compaction - because it's expensive operation
+// if last existing shard size merge it with delta
+// if serialized size of delta > ShardLimit - break down to multiple shards
+// shard number - it's biggest value in bitmap
 func AppendMergeByOr(c ethdb.Cursor, key []byte, delta *gocroaring.Bitmap) error {
-	shardNForDelta := uint16(0)
+	lastShardKey := make([]byte, len(key)+4)
+	copy(lastShardKey, key)
+	binary.BigEndian.PutUint32(lastShardKey[len(lastShardKey)-4:], ^uint32(0))
 
-	hotK, hotV, seekErr := c.Seek(key)
+	currentLastV, seekErr := c.SeekExact(lastShardKey)
 	if seekErr != nil {
 		return seekErr
 	}
-	if hotK != nil && bytes.HasPrefix(hotK, key) {
-		hotShardN := ^binary.BigEndian.Uint16(hotK[len(hotK)-2:])
-		if len(hotV) > int(HotShardLimit) { // merge hot to cold
-			var err error
-			shardNForDelta, err = hotShardOverflow(c, key, hotShardN, hotV)
-			if err != nil {
-				return err
-			}
-		} else { // merge delta to hot, write result to `hotShardN`
-			hot, err := gocroaring.Read(hotV)
-			if err != nil {
-				return err
-			}
 
-			delta = gocroaring.FastOr(delta, hot)
-			shardNForDelta = hotShardN
+	if currentLastV == nil { // no existing shards, then just create one
+		err := writeBitmapSharded(c, key, delta)
+		if err != nil {
+			return err
 		}
+		return nil
 	}
 
-	newK := make([]byte, len(key)+2)
-	copy(newK, key)
-	binary.BigEndian.PutUint16(newK[len(newK)-2:], ^shardNForDelta)
-
-	//delta.RunOptimize()
-	newV := make([]byte, delta.SerializedSizeInBytes())
-	err := delta.Write(newV)
+	last, err := gocroaring.Read(currentLastV)
 	if err != nil {
 		return err
 	}
-	err = c.Put(newK, newV)
-	if err != nil {
-		return err
-	}
-	return nil
-}
 
-func hotShardOverflow(c ethdb.Cursor, initialKey []byte, hotShardN uint16, hotV []byte) (shardNForDelta uint16, err error) {
-	if hotShardN == 0 { // no cold shards, create new hot from delta - it will turn hot to cold automatically
-		return 1, nil
-	}
+	if len(currentLastV) < int(ShardLimit) { // merge delta to last shard
```

Comment:
Can you simply do `delta = gocroaring.Or(delta, last)` and then `return writeBitmapSharded(c, key, delta)` without the rest of this function? As far as I can see, the `writeBitmapSharded` should take care of all of that